triggers/tests/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRuleSimulationLib.scala

```diff
@@ -6,93 +6,210 @@ package com.daml.lf.engine.trigger
 import akka.actor.ActorSystem
 import akka.stream.{FlowShape, KillSwitches, Materializer, SourceShape}
 import akka.stream.scaladsl.{Flow, GraphDSL, Keep, Sink, Source}
+import com.daml.ledger.api.refinements.ApiTypes.{ApplicationId, Party}
 import com.daml.ledger.api.v1.command_submission_service.SubmitRequest
 import com.daml.ledger.api.v1.event.CreatedEvent
+import com.daml.ledger.client.LedgerClient
+import com.daml.lf.PureCompiledPackages
 import com.daml.lf.data.FrontStack
+import com.daml.lf.data.Ref.{Identifier, PackageId, QualifiedName}
 import com.daml.lf.data.Time.Timestamp
 import com.daml.lf.engine.trigger.Runner.{TriggerContext, TriggerContextualFlow}
-import com.daml.lf.engine.trigger.ToLoggingContext._
 import com.daml.lf.engine.trigger.UnfoldState.{flatMapConcatNodeOps, toSourceOps}
-import com.daml.lf.speedy.SExpr.SEValue
-import com.daml.lf.speedy.SValue.{SList, SUnit}
-import com.daml.lf.speedy.{SValue, Speedy}
+import com.daml.lf.speedy.SValue.SList
+import com.daml.lf.speedy.SValue
 import com.daml.logging.LoggingContextOf
+import com.daml.platform.services.time.TimeProviderType
 import com.daml.util.Ctx
 
 import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+object TriggerRuleMetrics {
+  final case class EvaluationMetrics(
+      steps: Long,
+      submissions: Long,
+      getTimes: Long,
+      ruleEvaluation: FiniteDuration,
+      stepIteratorMean: FiniteDuration,
+      stepIteratorDelayMean: Option[FiniteDuration],
+  )
 
-class TriggerRuleSimulationLib[UserState](runner: Runner) {
+  final case class ACSMetrics(
+      activeContracts: Long,
+      pendingContracts: Long,
+  )
 
-  private[this] val loggingContext: LoggingContextOf[Trigger] =
-    LoggingContextOf.newLoggingContext(LoggingContextOf.label[Trigger])(identity)
+  final case class InFlightMetrics(
+      commands: Long
+  )
+
+  final case class SubmissionMetrics(
+      creates: Long,
+      exercises: Long,
+      createAndExercises: Long,
+      exerciseByKeys: Long,
+  )
+
+  final case class InternalStateMetrics(
+      acs: ACSMetrics,
+      inFlight: InFlightMetrics,
+  )
+
+  final case class RuleMetrics(
+      evaluation: EvaluationMetrics,
+      submission: SubmissionMetrics,
+      startState: InternalStateMetrics,
+      endState: InternalStateMetrics,
+  )
+}
+
+final class TriggerRuleSimulationLib private (
+    compiledPackages: PureCompiledPackages,
+    triggerId: Identifier,
+    triggerConfig: TriggerRunnerConfig,
+    client: LedgerClient,
+    timeProviderType: TimeProviderType,
+    applicationId: ApplicationId,
+    triggerParties: TriggerParties,
+) {
+
+  private[this] val logObserver: (String, TriggerLogContext) => Unit = { (_, _) =>
+    // TODO: a future PR will provide a more meaningful implementation
```

Comment:
Once the structured logging to trigger metric case class translation is defined, we can hook calls to the translation function here